src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -111,50 +143,46 @@ public bool TryResolveReferences(MetadataReferenceInfo metadataReferenceInfo, [N
+        }
 
-            if (_nameMap.TryGetValue(metadataReferenceInfo.FileName, out var _))
+        private void EnsureCachePopulated(string fileName)
+        {
+            if (!_nameToLocationsMap.TryGetValue(fileName, out var list))
             {
-                // The file name of this reference has already been searched for and none of them 
-                // had the correct MVID (else the _mvidMap lookup would succeed). No reason to do 
-                // more work here.
-                return false;
+                return;
             }
 
-            var list = new List<AssemblyInfo>();
+            _nameToLocationsMap.Remove(fileName);
 
-            foreach (var directory in _indexDirectories)
+            using var _ = _logger.BeginScope($"Populating {fileName}");
+            var assemblyInfoList = new List<AssemblyInfo>();
+            foreach (var filePath in list)
             {
-                foreach (var fileInfo in directory.EnumerateFiles(metadataReferenceInfo.FileName, SearchOption.AllDirectories))
+                if (Util.GetPortableExecutableInfo(filePath) is not { } peInfo)
                 {
-                    if (Util.GetPortableExecutableInfo(fileInfo.FullName) is not { } peInfo)
-                    {
-                        _logger.LogWarning($@"Could not read MVID from ""{fileInfo.FullName}""");
-                        continue;
-                    }
+                    _logger.LogWarning($@"Could not read MVID from ""{filePath}""");
+                    continue;
+                }
 
-                    if (peInfo.IsReadyToRun)
-                    {
-                        _logger.LogInformation($@"Skipping ReadyToRun image ""{fileInfo.FullName}""");
-                        continue;
-                    }
+                if (peInfo.IsReadyToRun)
+                {
+                    _logger.LogInformation($@"Skipping ReadyToRun image ""{filePath}""");
+                    continue;
+                }
 
-                    var currentInfo = new AssemblyInfo(fileInfo.FullName, peInfo.Mvid);
-                    list.Add(currentInfo);
+                var currentInfo = new AssemblyInfo(filePath, peInfo.Mvid);
+                assemblyInfoList.Add(currentInfo);
 
-                    if (!_mvidMap.ContainsKey(peInfo.Mvid))
-                    {
-                        _logger.LogTrace($"Caching [{peInfo.Mvid}, {fileInfo.FullName}]");
-                        _mvidMap[peInfo.Mvid] = currentInfo;
-                    }
+                if (!_mvidMap.ContainsKey(peInfo.Mvid))
+                {
+                    _logger.LogTrace($"Caching [{peInfo.Mvid}, {filePath}]");
+                    _mvidMap[peInfo.Mvid] = currentInfo;
                 }
             }
 
-            _nameMap[metadataReferenceInfo.FileName] = list;
-            return _mvidMap.TryGetValue(metadataReferenceInfo.ModuleVersionId, out assemblyInfo);
+            _nameMap[fileName] = assemblyInfoList;
```

Comment:
I guess the reason we don't do this at the same time we enumerate directories is, this step is more expensive and we don't expect to be asked about every last .exe or .dll? Is that right?

Reply:
That's correct. Enumerating the directories is expensive so we try to do it the minimum number of times. Essentially it's cheaper to enumerate `*.dll` once and record the locations vs. enumerate `fileName.dll` for each individual request. The enumeration itself is very expensive given the number of dirs that we search so do it once and save. 

But delay actually reading the files off disk until we're asked to do so cause we end up indexing a lot more files than we actually need to read.